pkg/controller/replicaset/replica_set.go

```diff
@@ -802,10 +804,28 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
 	if diff < len(filteredPods) {
 		podsWithRanks := getPodsRankedByRelatedPodsOnSameNode(filteredPods, relatedPods)
 		sort.Sort(podsWithRanks)
+		filteredPods = podsWithRanks.Pods
+		reportMetrics(filteredPods, diff)
 	}
 	return filteredPods[:diff]
 }
 
+func reportMetrics(filteredPods []*v1.Pod, diff int) {
+	youngestTime := time.Time{}
+	// first we need to check all of the pods to get the youngest, as they may not necessarily be sorted by timestamp alone
+	for _, pod := range filteredPods {
+		if time.Since(pod.CreationTimestamp.Time) < time.Since(youngestTime) {
+			youngestTime = pod.CreationTimestamp.Time
+		}
+	}
+
+	// for each pod chosen for deletion, report the ratio of its age to the youngest pod's age
+	for _, pod := range filteredPods[:diff] {
```

Comment:
In the KEP it says:

```
The metric `deleted_pod_age_ratio` will provide a histogram of the ratio between the 
chosen `deleted pod`'s age over the current `youngest pod`'s age, for pods where the sort 
algorithm falls back to age. (Pod age is the final criteria in the sorting algorithm, so we don't 
want to measure this ratio for deletions which don't use this feature, as those may validly fall 
outside the desired range).
```

However, it seems like it would be hard to tell. So I'm fine with this implementation.

Reply:
Yeah, I wanted to make that distinction but looking at the actual implementation we don't have an easy way to single out "what" caused a pod to be sorted in that order.

I do see now you pointed out the metric should be a histogram, so I'll update to that (I just used a gauge)

Reply:
So the problem is that currently we can have arbitrary values in the histogram in a valid way.
So that metric won't be useful by any means...

So let's try to make it useful:
- what we really care about are ready pods - so let's focus only on those
- so whenever we delete "ready pod" report its age compared to the youngest "ready" pod.

That solves most of the problems with the metric, modulo these two:
https://github.com/kubernetes/kubernetes/blob/4aae71695a8dd43918702fafd81e0401721d79d9/pkg/controller/controller_utils.go#L836-L851
- the first one we can ignore pods that set that annotation
- the former we can address by computing the youngest within the pods with the same rank
But that would make this metric very hard to intuitively explain to anyone. 

I didn't initially realize that the sorting function has so many different branches. And I'm not sure what to do. Up until we were ignoring non-ready pods, that metric still made some sense. But after that point it's becoming too artificial imho.

Any ideas what to do with this?

Reply:
I'm starting to doubt the usefulness of the metric as a whole. If we are removing several pods, it's likely that some of them will be more than twice older than the youngest.

Are we simply assuming that most of the times you would do small scale downs so the number should stay below 2?

If we want to be 100% accurate, we should be grouping by rank. But I think we would be missing an opportunity for the metric to be useful for users beyond checking that the feature works. For example, if the users see that the metric is well above 2, they can start wondering why, and the reason could be that they are not spreading their pods appropriately.

Reply:
This was the point I was trying to pass above.

But at the same time, I would really like to be able to answer if the feature works. But given all of the above, we should maybe be looking at the feature as a whole "sorting" function? Or maybe yet a different angle?
I'm looking for ideas what things we may try to report (probably in a wider context) to show that the feature works.

Reply:
So the sorting is trying to balance two goals/heuristics:
1. newer pods are generally safer to delete (hasn't warmed up, has done less work, has less open connections, etc).
2. pods should be spread. This is achieved in 2 ways:
  - deleting pods in nodes where there are other pods of the same replicaset
  - random deletion (the feature being graduated).

The proposed metric is trying to make sure we don't bomb 1. Perhaps we should keep it as it is already implemented in this PR.

Now, can we easily report a metric for 2 without increasing runtime significantly? The only information we have at this point is the node name. To make it really useful we should report the zone spread, but then we would have to add a node lister.

Reply:
@wojtek-t @alculquicondor in the interest of keeping this on track for code freeze, could I propose a compromise in that we continue with the metric as implemented in this PR, and add a TODO to continue investigating the other metric (possibly for GA promotion?). The conversation seems to have stalled on this one

Reply:
I was talking it over with @brancz, it may make more sense to have two metrics, from the gist of this conversation (without knowing too much about what you guys are actually doing), it sounds like you may want a `youngest_ready_pod_age` metric and a `deleted_ready_pod_age` metric, or something like this and do the computation after the fact.

Reply:
I don't think that would help. The ratio only makes sense at the time of deletion. If you have separate metrics, then you can't match the deleted pod with the "youngest" pod to do the division.

Reply:
I think we should be good with the metric as is, except it should be a distribution and only focus on Ready pods, as Wojtek proposed earlier. Any extra filtering sounds like an overkill to me.

Reply:
BTW, I'm proposing to get rid of the rank, in the context of #102949

Reply:
> The ratio only makes sense at the time of deletion. If you have separate metrics, then you can't match the deleted pod with the "youngest" pod to do the division.

This was my concern with the proposal for 2 metrics as well. Ultimately I think any measurement of this feature will have some level of awkwardness, that may require refinement. I agree that the metric as written is probably the simplest approach for now.

Reply:
You can use two Gauge Funcs, these will literally evaluate at scrape time, therefore will always expose only current data.

Reply:
Okay, I talked to @alculquicondor offline, I think I grok what the metric is actually trying to measure. The name is a bit confusing, I think I prefer `sorting_deletion_age_ratio`, since this metric primarily targets the efficacy of the sorting algorithm. I also retract my suggestion about the two metrics since I believe I understand the point of what this is measuring a bit better.

Reply:
@damemi, you missed the name change `sorting_deletion_age_ratio`

Reply:
thanks, updated